### ISSUE TITLE
throw error when before middleware returns undefined

### ISSUE THREAD

**Issue by capaj** [Huemul/trae#49]
I'd think this would be safer for most programmers who are like me and forget the return

**Comment by ianaya89** [comment]
Is not a bad idea, I think we can add a warning at least. @gillchristian @ndelvalle any thoughts?

**Comment by gillchristian** [comment]
Yup, I've been thinking about warnings. I don't know if `throw`ing is the way to go, but a warning surely is.

We could try something similar to what React does with `propTypes` checking only on development build.

**Comment by ndelvalle** [comment]
Yeah, I don't think that throwing is the way to go. A warning sounds good, but just for a dev build like @gillchristian suggested. The thing is that if we do it, we should do it for all functionalities, cause users that use this dev build will expect warnings for all functionalities used in a wrong way.

**Comment by ianaya89** [comment]
Makes sense, I like the idea, will improve devs experience and is a way to guide new devs how to use the lib. I would like to start working on this.

**Comment by gillchristian** [comment]
Awesome @ianaya89 !!!

**Comment by capaj** [comment]
@gillchristian is there a usecase when returning `undefined` is a valid value? Seemed to me there was not.

**Comment by gillchristian** [comment]
- `before`: expected (the last one*) to return the config for the request.
- `after`: is up to the user consuming the response. Maybe not even a warning necessary :thinking: 
- `finally`: does not matter because middlewares are just run one after the other, no params are passed, nor is expected for them to return something.

---

(*) Let's consider the next case:

```javascript
const noOp = () => {}

const overwriteConf  = () => ({
  headers: { 'Content-Type': 'application/json' }
  mode: 'no-cors'
})

trae.before(noOp)
trae.before(overwriteConf)
```

`noOp` runs first and sets the config object to `undefined`, `overwriteConf`gets that `undefined` "configuration object" but does nothing with it, instead it returns a proper configuration object.
I know this is a silly example, I don't imagine people doing stuff like this but it still shows that validation needs to be done to the returned value of the last middleware applied.

**Comment by gillchristian** [comment]
@ianaya89 check [this out](https://github.com/facebook/react/blob/c78464f8ea9a5b00ec80252d20a71a1482210e57/src/shared/utils/flattenChildren.js#L21-L25).

**Comment by ndelvalle** [comment]
It would be nice if we dont even have statements like the following in the prod bundle.
```javascript
if (process.env.NODE_ENV !== 'production') {
  // check and do some warnings 
}
```
